scripts/run_inference.py

```diff
@@ -173,14 +178,24 @@ def load(
         try:
             with AdapterConfigPatcher(checkpoint_path, overrides):
                 try:
-                    model = AutoPeftModelForCausalLM.from_pretrained(checkpoint_path)
+                    model = AutoPeftModelForCausalLM.from_pretrained(
+                        checkpoint_path,
+                        attn_implementation="flash_attention_2"
+                        if use_flash_attn
+                        else None,
+                        torch_dtype=torch.float16 if use_flash_attn else None,
```

Comment:
Is there a reason that float16 is being hardcoded here since flash attention also supports bfloat32?

Reply:
Yeah I wasn't sure about this, I think I did it based on [this comment](https://github.com/foundation-model-stack/fms-hf-tuning/issues/103#issuecomment-2036003946) and perhaps some examples I saw like [here](https://github.com/huggingface/transformers/blob/bc274a28a9a72c5fe98b070e2083fe689c8fd5e9/docs/source/en/perf_infer_gpu_one.md?plain=1#L102-L115). But I think you're very right that this should be removed

Reply:
If it's removed, what torch_dtype will be used to load the model? If trying to load a model that uses float32 for example this would fail so should this be set by the user? and default to bfloat16 as we have in [sft_trainer](https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/tuning/sft_trainer.py#L134-L135)?

For example I get warning:
```
Flash Attention 2.0 only supports torch.float16 and torch.bfloat16 dtypes, but the current dype in GPTBigCodeModel is torch.float32. You should run training or inference using Automatic Mixed-Precision via the `with torch.autocast(device_type='torch_device'):` decorator, or load the model with the `torch_dtype` argument. Example: `model = AutoModel.from_pretrained("openai/whisper-tiny", attn_implementation="flash_attention_2", torch_dtype=torch.float16)`
```
and then error `RuntimeError: FlashAttention only support fp16 and bf16 data type`

Reply:
Ahh sorry, I meant bfloat16, not bfloat32 :) The best choice probably depends on what type the model is tuned with, but the safer choice is probably `bfloat16` since it has the a bigger exponent (same range as float32).

It would be nice to check the behavior for what happens if you try loading with bfloat16 on a device that doesn't support it, to see if it has a fallback (and if that actually modifies the loaded model `dtype` value, or just the way computations are handled) or throws. But the most reasonable choice is probably to do the same thing we are doing for tuning to select the default data type at tune time, or check if bf16 is supported and use it if we can. The easiest way to check if bf16 is available is:

`torch.cuda.is_available() and torch.cuda.is_bf16_supported()`
(the latter throws if you try to call it in torch with no cuda installed)

Reply:
We use `bfloat16` as the default for tuning ([ref](https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/tuning/config/configs.py#L40)). So `bfloat16` is the right default here, and I assume since we don't explicitly handle it, `from_pretrained` has a sensible fallback

Reply:
Modified to load with bfloat16 and I tried testing on a v100 where `torch.cuda.is_available() == True` and `torch.cuda.is_bf16_supported() == False`. How do I check the dtype of the model? I ran...
```py
>>> model = AutoModelForCausalLM.from_pretrained("bigscience/bloom-560m", torch_dtype=torch.bfloat16)
config.json: 100%|█████████████████████████████████████████████████████████████████████████| 693/693 [00:00<00:00, 4.49MB/s]
model.safetensors: 100%|███████████████████████████████████████████████████████████████| 1.12G/1.12G [00:14<00:00, 79.3MB/s]
>>> model.config.torch_dtype
torch.bfloat16
```

Reply:
The `config.torch_dtype` is set by the model at save time, I think! The one you should want is just the model `.dtype`, which reflects the current instance loaded on your device, I don't think conversions like `.to` (which moves the model to different devices/types etc)  would update the config based attribute but I am not sure